sled-agent/src/storage/dump_setup.rs

```diff
@@ -52,103 +116,458 @@ impl DumpSetup {
+
+                            Self::copy_sync_and_remove(&entry.path(), &dest)?;
+
+                            info!(self.log, "Relocated {entry:?} to {dest:?}");
+                        } else {
+                            error!(self.log, "Non-UTF8 path found while rotating core dumps: {entry:?}");
                         }
                     }
                 }
             }
+        } else {
+            info!(
+                self.log,
+                "No rotation destination for crash dumps yet chosen."
+            );
+        }
+
+        if let Err(err) = self.rotate_logs() {
+            error!(
+                self.log,
+                "Failure while trying to rotate logs to debug dataset: {err:?}"
+            );
+        }
+
+        Ok(())
+    }
+
+    fn copy_sync_and_remove(
+        source: impl AsRef<Path>,
+        dest: impl AsRef<Path>,
+    ) -> std::io::Result<()> {
+        let source = source.as_ref();
+        let dest = dest.as_ref();
+        let mut dest_f = std::fs::File::create(&dest)?;
+        let mut src_f = std::fs::File::open(&source)?;
+
+        std::io::copy(&mut src_f, &mut dest_f)?;
+
+        dest_f.sync_all()?;
+
+        drop(src_f);
+        drop(dest_f);
+
+        std::fs::remove_file(source)?;
+        Ok(())
+    }
+
+    fn rotate_logs(&self) -> Result<(), RotateLogsError> {
+        let debug_dir = self
+            .chosen_debug_dir
+            .as_ref()
+            .ok_or(RotateLogsError::NoDebugDirYet)?;
+        // zone crate's 'deprecated' functions collide if you try to enable
+        // its 'sync' and 'async' features simultaneously :(
+        let rt =
+            tokio::runtime::Runtime::new().map_err(RotateLogsError::Tokio)?;
+        let oxz_zones = rt.block_on(Zones::get())?;
+        Self::rotate_logs_inner(
+            debug_dir,
+            PathBuf::from("/var/svc/log"),
+            "global",
+        )?;
+        for zone in oxz_zones {
+            let logdir = zone.path().join("root/var/svc/log");
+            let zone_name = zone.name();
+            Self::rotate_logs_inner(debug_dir, logdir, zone_name)?;
+        }
+        Ok(())
+    }
+
+    fn rotate_logs_inner(
+        debug_dir: &DebugDirPath,
+        logdir: PathBuf,
+        zone_name: &str,
+    ) -> Result<(), RotateLogsError> {
+        // pattern matching rotated logs, e.g. foo.log.3
+        let pattern = logdir
+            .join("*.log.*")
+            .to_str()
+            .ok_or_else(|| RotateLogsError::Utf8(zone_name.to_string()))?
+            .to_string();
+        let glob = glob::glob(&pattern)?;
+        for entry in glob.flatten() {
+            let dest_dir = debug_dir.join(zone_name).into_std_path_buf();
+            std::fs::create_dir_all(&dest_dir)?;
+            let src_name = entry.file_name().unwrap();
+            // as we rotate them out, logadm will keep resetting to .log.0,
+            // so we need to maintain our own numbering in the dest dataset
```

Comment:
I think we should reconfigure logadm in the zone to rotate the files using a date stamp pattern, like `2023-07-19T10:00:00Z` (`%FT%TZ` maybe?) rather than try to work around the integer suffixes.